@pytest.mark.trio crash with pytest 3.8.1

pytest 3.8.1 seems to be incompatible with our use of `@pytest.mark.trio` marking

Exemple:
```python
import pytest

class TestFoo:

    @pytest.mark.trio
    async def test_bar(self):
        pass
```

With pytest 3.8.1
```
$ pip freeze | grep pytest
pytest==3.8.1
pytest-cov==2.6.0
pytest-logbook==1.2.0
pytest-trio==0.5.0
$ py.test tests/test_tt.py   -vvv
============================================================================== test session starts ==============================================================================
platform linux -- Python 3.6.1, pytest-3.8.1, py-1.6.0, pluggy-0.7.1 -- /home/emmanuel/projects/parsec-cloud/venv2/bin/python3.6
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/emmanuel/projects/parsec-cloud/.hypothesis/examples')
rootdir: /home/emmanuel/projects/parsec-cloud, inifile: setup.cfg
plugins: trio-0.5.0, logbook-1.2.0, cov-2.6.0, hypothesis-3.71.10
collected 1 item                                                                                                                                                                

tests/test_tt.py::TestFoo::test_bar FAILED                                                                                                                                [100%]

=================================================================================== FAILURES ====================================================================================
_______________________________________________________________________________ TestFoo.test_bar ________________________________________________________________________________

value = <trio._core._run.Nursery object at 0x7f502b404d30>

    async def yield_(value=None):
>       return await _yield_(value)

venv2/lib/python3.6/site-packages/async_generator/_impl.py:106: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

value = <trio._core._run.Nursery object at 0x7f502b404d30>

    @coroutine
    def _yield_(value):
>       return (yield _wrap(value))
E       TypeError: test_bar() missing 1 required positional argument: 'self'

venv2/lib/python3.6/site-packages/async_generator/_impl.py:99: TypeError
=========================================================================== 1 failed in 0.09 seconds ============================================================================
```

And with good old pytest 3.8 :smile: 
```
$ pip freeze | grep pytest
pytest==3.8.0
pytest-cov==2.6.0
pytest-logbook==1.2.0
pytest-trio==0.5.0
$ py.test tests/test_tt.py   -vvv
============================================================================== test session starts ==============================================================================
platform linux -- Python 3.6.1, pytest-3.8.0, py-1.6.0, pluggy-0.7.1 -- /home/emmanuel/projects/parsec-cloud/venv2/bin/python3.6
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/emmanuel/projects/parsec-cloud/.hypothesis/examples')
rootdir: /home/emmanuel/projects/parsec-cloud, inifile: setup.cfg
plugins: trio-0.5.0, logbook-1.2.0, cov-2.6.0, hypothesis-3.71.10
collected 1 item                                                                                                                                                                

tests/test_tt.py::TestFoo::test_bar PASSED                                                                                                                                [100%]

=========================================================================== 1 passed in 0.01 seconds ============================================================================
```

--- a/newsfragments/64.bugfix.rst
+++ b/newsfragments/64.bugfix.rst
@@ -1,0 +1,2 @@
+The pytest 3.8.1 release broke pytest-trio's handling of trio tests
+defined as class methods. We fixed it again.

--- a/pytest_trio/_tests/conftest.py
+++ b/pytest_trio/_tests/conftest.py
@@ -3,7 +3,8 @@
 import pytest
 import warnings
 warnings.filterwarnings(
-    "default", category=pytest.RemovedInPytest4Warning,
+    "default",
+    category=pytest.RemovedInPytest4Warning,
     message=".*non-top-level conftest.*",
 )
 

--- a/pytest_trio/_tests/conftest.py
+++ b/pytest_trio/_tests/conftest.py
@@ -1,1 +1,10 @@
+# Temporary hack while waiting for an answer here:
+#    https://github.com/pytest-dev/pytest/issues/4039
+import pytest
+import warnings
+warnings.filterwarnings(
+    "default", category=pytest.RemovedInPytest4Warning,
+    message=".*non-top-level conftest.*",
+)
+
 pytest_plugins = ["pytester"]

--- a/pytest_trio/_tests/test_trio_asyncio.py
+++ b/pytest_trio/_tests/test_trio_asyncio.py
@@ -9,8 +9,8 @@ except ImportError:
     pytestmark = pytest.mark.skip(reason="trio-asyncio not available")
 
 
-async def use_run_asyncio():
-    await trio_asyncio.run_asyncio(asyncio.sleep, 0)
+async def use_asyncio():
+    await trio_asyncio.aio_as_trio(asyncio.sleep)(0)
 
 
 @pytest.fixture()
@@ -23,7 +23,7 @@ async def asyncio_loop():
 @pytest.fixture()
 @async_generator
 async def asyncio_fixture_with_fixtured_loop(asyncio_loop):
-    await use_run_asyncio()
+    await use_asyncio()
     await yield_()
 
 
@@ -31,21 +31,21 @@ async def asyncio_fixture_with_fixtured_loop(asyncio_loop):
 @async_generator
 async def asyncio_fixture_own_loop():
     async with trio_asyncio.open_loop():
-        await use_run_asyncio()
+        await use_asyncio()
         await yield_()
 
 
 @pytest.mark.trio
 async def test_no_fixture():
     async with trio_asyncio.open_loop():
-        await use_run_asyncio()
+        await use_asyncio()
 
 
 @pytest.mark.trio
 async def test_half_fixtured_asyncpg_conn(asyncio_fixture_own_loop):
-    await use_run_asyncio()
+    await use_asyncio()
 
 
 @pytest.mark.trio
 async def test_fixtured_asyncpg_conn(asyncio_fixture_with_fixtured_loop):
-    await use_run_asyncio()
+    await use_asyncio()

--- a/pytest_trio/plugin.py
+++ b/pytest_trio/plugin.py
@@ -292,7 +292,7 @@ class TrioFixture:
 
 
 def _trio_test_runner_factory(item, testfunc=None):
-    testfunc = testfunc or item.function
+    testfunc = testfunc or item.obj
 
     if getattr(testfunc, '_trio_test_runner_wrapped', False):
         # We have already wrapped this, perhaps because we combined Hypothesis

